### ISSUE TITLE
Add ``zip_with_scalars()`` function

### ISSUE THREAD

**Issue by kalekundert** [more-itertools/more-itertools#543]
### Description
Would you be interested in adding a function that mostly works like `zip()`, except that it treats any non-iterable arguments as values that should be included in each output tuple (without affecting how the iterable arguments are zipped)?

### References
I needed a function like this when I was writing a function for reading keys from JSON files.  The most common way to call the function would be with one key and one file, but it was also important to allow multiple keys or files to be specified.  For example, consider these two files:
```
# foo.json
{"a": 1, "b": 2}
```
```
# bar.json
{"a": 3, "c": 4}
```
I needed to be able to do:

    >>> f("foo.json", "a")
    [1]
    >>> f("foo.json", ["a", "b"])
    [1, 2]
    >>> f(["foo.json", "bar.json"], "a")
    [1, 3]
    >>> f(["foo.json", "bar.json"], ["b", "c"])
    [2, 4]

This is pretty hard to write in a general way.  You can't just wrap the scalars in `itertools.repeat`, because you'll get an infinite loop if only scalars are specified.

### Examples
Here are some examples to show how it works:
```
>>> list(zip_with_scalars(1, 2))
[(1, 2)]
>>> list(zip_with_scalars(1, [2]))
[(1, 2)]
>>> list(zip_with_scalars(1, [2, 3]))
[(1, 2), (1, 3)]
>>> list(zip_with_scalars([1, 2], [3, 4]))
[(1, 3), (2, 4)]
```

Some special cases:
```
>>> list(zip_with_scalars("ab"))
[("ab",)]
>>> list(zip_with_scalars("ab", not_iterable=None))
[("a", "b")]
>>> list(zip_with_scalars([1], [2, 3], strict=True))
Traceback...
UnequalIterableError
```

### Implementation
Here's an implementation.  If you're interested in adding this itertool, I can turn this into a PR with tests and docs:
```
def zip_with_scalars(*objs, strict=False, not_iterable=(str, bytes)):
    from builtins import zip
    from operator import itemgetter
    from more_itertools import zip_equal

    def is_scalar(obj):
        if not_iterable and isinstance(obj, not_iterable):
            return True
        try:
            iter(obj)
        except TypeError:
            return True
        else:
            return False

    iterables = []
    formatters = []

    for obj in objs:
        if is_scalar(obj):
            # The double-lambdas are necessary to create a closure.
            formatters.append((lambda x: lambda _: x)(obj))
        else:
            formatters.append(itemgetter(len(iterables)))
            iterables.append(obj)

    if not iterables:
        if not objs:
            return
        else:
            yield tuple(objs)
            return

    if strict:
        zip = zip_equal

    for values in zip(*iterables):
        yield tuple(f(values) for f in formatters)
```

**Comment by bbayles** [comment]
Thanks for the writeup. I'm inclined to pass on this one, but if you see places in the wild (StackOverflow answers, GitHub code aimed at the same problem) do let me know.

**Comment by kalekundert** [comment]
Ok, will do.

**Comment by kalekundert** [comment]
Here are some stack overflow posts that are basically asking for this algorithm:

- https://stackoverflow.com/questions/26311561/zip-list-of-scalars-and-lists
- https://stackoverflow.com/questions/22743490/zip-in-python-how-to-use-static-values
- https://stackoverflow.com/questions/36536171/zip-list-with-a-single-element

**Comment by bbayles** [comment]
OK, thanks for that. Yes, let's do a PR for this.

If the only reason we can't do `repeat` is the "all scalar" case, why don't we just detect that? We should have the ability.

```python
def zip_fill(*objects, scalar_types=(str, bytes), **kwargs):
    iterables = []
    all_scalar = True
    for obj in objects:
        if not_iterable and isinstance(obj, scalar_types):
            iterables.append(repeat(obj))
        else:
            try:
                iterables.append(iter(obj))
            except TypeError:
                iterables.append(repeat(obj))
            else:
                all_scalar = False

    if all_scalar:
        yield from zip(*([obj] for obj in objects))
    else:
        yield from zip(*iterables, **kwargs)
```

**Comment by kalekundert** [comment]
I think using `repeat()` is also incompatible with `zip(strict=True)`, since the iterables won't have the same length.  Note that the `strict` argument will be [new in python 3.10](https://docs.python.org/3.10/library/functions.html#zip), but I think it's a good API and worth pre-emptively matching.

**Comment by bbayles** [comment]
Could you show why it's not compatible? In my sketch above the `all_scalar` case doesn't pass on `**kwargs`, because `strict=True` is irrelevant - all iterables are the same size.

In the "not `all_scalar`" case, it does pass on `**kwargs` (including `strict`). Shouldn't this raise an exception, since the iterables aren't all the same size?

**Comment by kalekundert** [comment]
I see what you're getting at, but I think that the `strict` argument should only apply to the iterables, e.g.:
```
>>> list(zip_fill(1, [2, 3], strict=True))
[(1, 2), (1, 3)]
>>> list(zip_fill(1, [2, 3], [4,5,6], strict=True))
Traceback...
```
My reasoning is that scalars behave like iterables of the same length as the shortest specified iterable argument (or 1 if there are no iterable arguments).  This implies that they shouldn't have any effect on the `strict` check.  I'll also add that it doesn't make sense to think of scalars as just being "iterables with length 1", because they behave differently:
```
>>> list(zip_fill(1, [2, 3]))
[(1, 2), (1, 3)]
>>> list(zip_fill([1], [2, 3]))
[(1, 2)]
```

Another thing worth pointing out is that `zip_fill(strict=True)` is pretty similar to [`np.broadcast`](https://numpy.org/doc/stable/user/basics.broadcasting.html) (see [this](https://stackoverflow.com/questions/15167615/is-there-a-good-way-of-iterating-through-several-listswhere-sometimes-they-are) Stack Overflow post), which has the behavior I'm proposing:
```
>>> list(np.broadcast(1, [2,3]))
[(1, 2), (1, 3)]
>>> list(np.broadcast(1, [2,3], [4,5,6]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: shape mismatch: objects cannot be broadcast to a single shape
```

**Comment by kalekundert** [comment]
Thanks for taking the time to discuss this.  I know it's a tedious detail, but it's important to get right.

The "1" and "infinite" algorithms have a certain conceptual purity, but they just don't make sense when applied to the actual problems this function is supposed to solve.

To explain in more depth, here's how I understand the `strict` argument:
- The purpose of `strict=True` is to prevent values from being silently ignored.
- Best practice (as outlined in the [python 3.10 docs](https://docs.python.org/3.10/library/functions.html#zip)) is to specify `strict=True` unless you actively intend to ignore values.
- The `strict` argument needs to be part of the API; it's unsafe to use `zip` without it.

With that in mind, let's look at all of the examples I've brought up so far in this thread.  Every single one is incompatible with the "1" and "infinite" algorithms.  The "shortest" and "longest" algorithms are indistinguishable for the version of the function I'm proposing, but you're right that "longest" is more correct if you allow for the possibility of implementing something like `zip_longest()` (which I would be open to).

- https://stackoverflow.com/questions/26311561/zip-list-of-scalars-and-lists:
  - Input: 
    ```
    (25.69, 
     [25.5, 29.5],
     [0.00153, 0.0222],
     0.0, 
     [0.04109, 0.08877],
     [0.31393, 0.34876],
     [0.75, 0.63])
    ```
  - Desired output:
    ```
    ((25.69, 25.5, 0.00153, 0.0, 0.04109, 0.31393, 0.75),
     (25.69, 29.5, 0.222, 0.0, 0.08877, 0.34876, 0.63))
    ```
  - The desired output includes all input values; no values are ignored.
  - That means the `strict` argument should be `True`.
  - The "1" and "infinite" algorithms would raise.
  - The "longest" algorithm would return the desired output.

- https://stackoverflow.com/questions/22743490/zip-in-python-how-to-use-static-values:
  - Input:
    ```
    [1,2,3,4,5], 100, 200
    ```
  - Desired output: 
    ```
    [(1,100,200),
     (2,100,200),
     (3,100,200),
     (4,100,200),
     (5,100,200)]
    ```
  - The desired output includes all input values; no values are ignored.
  - That means the `strict` argument should be `True`.
  - The "1" and "infinite" algorithms would raise.
  - The "longest" algorithm would return the desired output.

- https://stackoverflow.com/questions/36536171/zip-list-with-a-single-element

  - Input:
    ```
    [1, 2, 3, 4], 'a'
    ```
  - Desired output:
    ```
    [(1, 'a'), (2, 'a'), (3, 'a'), (4, 'a')]
    ```
  - The desired output includes all input values; no values are ignored.
  - That means the `strict` argument should be `True`.
  - The "1" and "infinite" algorithms would raise.
  - The "longest" algorithm would return the desired output.

- My own JSON example:
  - The actual code I wrote is on github: [`load_parameters()`](https://github.com/kalekundert/parametrize_from_file/blob/9eef273505d7202573fbbc7def8c5c647387c837/parametrize_from_file/parametrize.py#L220)
  - Example usage (modified from above, albeit still simplified):
    ```
    >>> load_parameters("foo.json", "a")
    [1]
    >>> load_parameters("foo.json", ["a", "b"])
    [1, 2]
    >>> load_parameters(["foo.json", "bar.json"], "a")
    [1, 3]
    >>> load_parameters(["foo.json", "bar.json"], ["b", "c"])
    [2, 4]
    ```
  - This code actually uses the `strict` argument to generate a good error message if iterables of different lengths are specified.
  - The "1" and "infinite" algorithms would raise for 2 of the 4 examples and would be unable to check for errors.
  - The "longest" algorithm returns the desired output in all 4 cases and correctly checks for errors.


In other words, the "1" and "infinite" algorithms make `strict=True` functionally useless.  I'd even go so far as to say that they make the whole function useless, because it's dangerous to use `zip` without `strict=True`.

**Comment by kalekundert** [comment]
> Perhaps we should just go forward without considering strict? The numpy function doesn't seem to have it.

`np.broadcast()` *only* has `strict=True`.  It doesn't have the option to silently ignore values from iterables of different lengths, and its behavior matches what I've been proposing for `strict=True`.  

I'm strongly against going forward without the `strict` argument.  That argument is part of the `zip()` API, so a `more_itertools` function based on `zip()` should implement it.

**Comment by bbayles** [comment]
I presume PR https://github.com/more-itertools/more-itertools/pull/545 does what you want? Please do leave comments there.
